src/liballoc/collections/btree/set.rs

```diff
@@ -109,6 +109,80 @@ pub struct Range<'a, T: 'a> {
     iter: btree_map::Range<'a, T, ()>,
 }
 
+/// Core of an aLternative to Peekable, that is more efficient
+/// if copying items is cheap and if peeking early does no harm.
+/// It always has one item read ahead.
+#[derive(Clone)]
+struct Peeking<I>
+where
+    I: Iterator,
+    I::Item: Copy,
+{
+    head: Option<I::Item>,
+    tail: I,
+}
+impl<I> Peeking<I>
+where
+    I: ExactSizeIterator,
+    I::Item: Copy,
+{
+    fn new(mut iter: I) -> Self {
+        let head = iter.next();
+        Peeking { head, tail: iter }
+    }
+    fn next(&mut self) -> Option<I::Item> {
+        match self.head {
```

Comment:
Is this match necessary for correctness? I'd say it's not, since all set iterators should be fused already. So it could be removed if that's better for performance, and you'd have just this?

```rust
item = self.head;
self.head = self.tail.next();
item
```

Not sure if it's a win, at all.

Reply:
For correctness, requiring FusedIterator (like I unintentionally left it in #65226) seems solid to me. But relying on that is bad for the performance I measure. That's also why the enum in #65226 is `A(Option<I::Item>)` and not `A(I::Item)`.

I don't quite understand why. A philosophical argument might me: a FusedIterator indeed keeps returning None, but nobody said it has to do so efficiently. But as far as I can tell, the map Keys iterator can't be any quicker: it just zero-compares the length that is red hot in the cache after updating it every iteration.

Maybe it's because I'm developing and benching this in a separate crate, which means the map Keys code cannot be inlined, and I'm mostly measuring (cross-DLL?) function calls?

Reply:
The keys code can inlined (it is generic, instantiated into the crate where it is used). One might tweak down to codegen-units=1 to experiment if that make a difference.

If it doesn't help, we'll just not do that change :)

Reply:
Oops, I remembered wrong. Your shortcut helps and it helped in earlier versions. Just don't use a `mem:replace` oneliner instead, that's way slower.